### ISSUE TITLE
need help setting up for a multiplatform project

### ISSUE THREAD

**Issue by Lakedaemon** [Strumenta/antlr-kotlin#33]
I could not find version 0.0.6
so I used 
val antlrKotlinVersion = "86a86f1968"

but with kotlin("multiplatform") version "1.3.61" (or "1.3.50")

Because Gradle (5.6.2) doesn't seem to recognize 
com.strumenta.antlrkotlin.gradleplugin.AntlrKotlinTask

this part cannot compile :

tasks.register<com.strumenta.antlrkotlin.gradleplugin.AntlrKotlinTask>("generateKotlinCommonGrammarSource") {
    // the classpath used to run antlr code generation
    antlrClasspath = configurations.detachedConfiguration(
        // antlr itself
        // antlr is transitive added by antlr-kotlin-target,
        // add another dependency if you want to choose another antlr4 version (not recommended)
        // project.dependencies.create("org.antlr:antlr4:$antlrVersion"),

        // antlr target, required to create kotlin code
        project.dependencies.create("com.strumenta.antlr-kotlin:antlr-kotlin-target:$antlrKotlinVersion")
    )
    maxHeapSize = "64m"
    packageName = "com.strumenta.antlrkotlin.examples"
    arguments = listOf("-no-visitor", "-no-listener")
    source = project.objects
        .sourceDirectorySet("antlr", "antlr")
        .srcDir("src/commonAntlr/antlr").apply {
            include("*.g4")
        }
    // outputDirectory is required, put it into the build directory
    // if you do not want to add the generated sources to version control
    outputDirectory = File("build/generated-src/commonAntlr/kotlin")
    // use this settings if you want to add the generated sources to version control
    // outputDirectory = File("src/commonAntlr/kotlin")
}

where did I err ?

my complete build file : 
import org.jetbrains.kotlin.resolve.calls.model.ResolvedCallArgument.DefaultArgument.arguments

val kotlinx_io_version = "0.1.13"
val kotlinx_coroutines = "1.3.3"
val spek_version = "2.0.6"
val antlrKotlinVersion = "86a86f1968"
buildscript {
    val kotlin_version = "1.3.61"
    val antlrKotlinVersion = "86a86f1968"

//    val kotlinx_io_version = "0.1.13"

    repositories {
        jcenter()
        maven(
            url ="https://jitpack.io"
        )
    }

    dependencies {
        classpath("org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version")
        classpath("org.jetbrains.kotlin:kotlin-serialization:$kotlin_version")
        classpath("com.strumenta.antlr-kotlin:antlr-kotlin-gradle-plugin:$antlrKotlinVersion")
    }
}

val group = "org.lakedaemon"
val version = "0.0.1"
plugins {
    kotlin("multiplatform") version "1.3.61"
    // id("kotlinx-serialization").version ("0.14.0")
    id("org.jetbrains.kotlin.plugin.serialization").version("1.3.61")
}
//apply (plugin= "kotlin-multiplatform")
//apply (plugin= "kotlinx-serialization")

repositories {
    // used for local development and while building by travis ci and jitpack.io
    mavenLocal()
    // used to download antlr4
    mavenCentral()
    // used to download korio
    jcenter()
    // used to download antlr-kotlin-runtime
    maven(url = "https://jitpack.io")
    // kotlin serialization
    maven(url = "https://kotlin.bintray.com/kotlinx")
    // spek
    maven(url = "https://dl.bintray.com/spekframework/spek-dev")
}

// Minimum jvmTarget of 1.8 needed since Kotlin 1.1
/*compileKotlin {
    kotlinOptions.jvmTarget= 1.8
}*/

//apply plugin: "maven-publish"

kotlin {
    targets {
        jvm()
        js()
        //fromPreset(presets.jvm, "jvm")
        //fromPreset(presets.js, "js")
        // For ARM, preset should be changed to presets.iosArm32 or presets.iosArm64
        // For Linux, preset should be changed to e.g. presets.linuxX64
        // For MacOS, preset should be changed to e.g. presets.macosX64
        //fromPreset(presets.linuxX64, "linux")
    }

    sourceSets {
        val commonAntlr by creating {
            dependencies {
                api(kotlin("stdlib-common"))
                // add antlr-kotlin-runtime
                // otherwise, the generated sources will not compile
                api("com.strumenta.antlr-kotlin:antlr-kotlin-runtime:$antlrKotlinVersion")
                // antlr-kotlin-runtime-jvm is automatically added as an jvm dependency by gradle
            }
            // you have to add the generated sources the to the kotlin compiler source directory list
            // this is not required if you use src/commonAntlr/kotlin
            // and want to add the generated sources to version control
            kotlin.srcDir("build/generated-src/commonAntlr/kotlin")
        }

        val commonMain by getting {
            dependsOn(commonAntlr)
            dependencies {
                implementation("org.jetbrains.kotlin:kotlin-stdlib-common")
                implementation("org.jetbrains.kotlinx:kotlinx-serialization-runtime-common:0.10.0")
                implementation("org.jetbrains.kotlinx:kotlinx-io-jvm:$kotlinx_io_version")
                implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core-common:$kotlinx_coroutines")
            }
        }
        val commonTest by getting {
            dependencies {
                implementation("org.jetbrains.kotlin:kotlin-test-common")
                // implementation ("org.jetbrains.kotlin:kotlin-test-annotations-common"

                implementation("org.spekframework.spek2:spek-dsl-metadata:$spek_version")
            }
        }
        val jvmMain by getting {

            dependencies {
                // antlr-kotlin
                api(kotlin("stdlib-jdk8"))
                api(kotlin("reflect"))

                implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk8")
                implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.3.3")
                implementation("org.jetbrains.kotlinx:kotlinx-coroutines-javafx:1.3.3")

                implementation("org.antlr:antlr4-runtime:4.7.1")
                //implementation (fileTree(dir= "libs", include= "*.jar"))

                implementation("org.jetbrains.kotlinx:kotlinx-serialization-runtime:0.10.0")

                /** for Mephistolus */
                implementation("no.tornado:tornadofx:1.7.19")

                /** we only need this for converting metamath ro mephistolus engine */
                implementation("org.apache.tinkerpop:gremlin-core:3.3.0")
                implementation("org.janusgraph:janusgraph-core:0.2.0")
                implementation("org.janusgraph:janusgraph-lucene:0.2.0")
                implementation("org.janusgraph:janusgraph-berkeleyje:0.2.0")
            }
        }
        val jvmTest by getting {
            dependencies {
                // antlr-kotlin
                api(kotlin("test-junit5"))

                implementation("org.jetbrains.kotlin:kotlin-test")
                // implementation ("org.jetbrains.kotlin:kotlin-test-junit"

                implementation("org.spekframework.spek2:spek-dsl-jvm:$spek_version")

                runtimeOnly("org.spekframework.spek2:spek-runner-junit5:$spek_version")
                runtimeOnly("org.jetbrains.kotlin:kotlin-reflect")
                implementation("org.jetbrains.kotlinx:kotlinx-serialization-runtime:0.10.0")
            }
        }
        val jsMain by getting {
            dependencies {
                implementation("org.jetbrains.kotlin:kotlin-stdlib-js")
            }
        }
        val jsTest by getting {
            dependencies {
                implementation("org.jetbrains.kotlin:kotlin-test-js")
            }
        }
        /*linuxMain {
        }
        linuxTest {
        }*/
    }
}

tasks {
    withType<Test> {
        useJUnitPlatform {
            includeEngines("spek2")
        }
    }
    withType<org.jetbrains.kotlin.gradle.tasks.KotlinCompile>().all {
        kotlinOptions {
            jvmTarget = "1.8"
        }
    }
}

// in antlr-kotlin-plugin <0.0.5, the configuration was applied by the plugin.
// starting from verison 0.0.5, you have to apply it manually:
tasks.register<com.strumenta.antlrkotlin.gradleplugin.AntlrKotlinTask>("generateKotlinCommonGrammarSource") {
    // the classpath used to run antlr code generation
    antlrClasspath = configurations.detachedConfiguration(
        // antlr itself
        // antlr is transitive added by antlr-kotlin-target,
        // add another dependency if you want to choose another antlr4 version (not recommended)
        // project.dependencies.create("org.antlr:antlr4:$antlrVersion"),

        // antlr target, required to create kotlin code
        project.dependencies.create("com.strumenta.antlr-kotlin:antlr-kotlin-target:$antlrKotlinVersion")
    )
    maxHeapSize = "64m"
    packageName = "com.strumenta.antlrkotlin.examples"
    arguments = listOf("-no-visitor", "-no-listener")
    source = project.objects
        .sourceDirectorySet("antlr", "antlr")
        .srcDir("src/commonAntlr/antlr").apply {
            include("*.g4")
        }
    // outputDirectory is required, put it into the build directory
    // if you do not want to add the generated sources to version control
    outputDirectory = File("build/generated-src/commonAntlr/kotlin")
    // use this settings if you want to add the generated sources to version control
    // outputDirectory = File("src/commonAntlr/kotlin")
}

// run generate task before build
// not required if you add the generated sources to version control
// you can call the task manually in this case to update the generated sources
tasks.getByName("compileKotlinJvm").dependsOn("generateKotlinCommonGrammarSource")

/*jvmMephistolusMain {
kotlinOptions {
    // Setup the Kotlin compiler options for the "main" compilation:
    jvmTarget = "1.8"
}
}*/

**Comment by drieks** [comment]
Hi,

sorry, but it is not possible to copy&paste your file into a local file because the content is not displayed correctly. Please try to edit your post (or just add a new comment) and write \`\`\` followed by a newline, then the file content and then \`\`\` again after the last line. You can also paste the content into the editor, select the content and then click the `<>` Icon in the Toolbar above the editor, this will add the three "grave accent" around the selected text.

Do you have a `settings.gradle.kts` in your project? if not, please try to add this one:
```
pluginManagement {
    repositories {
        // used for local development and while building by travis ci and jitpack.io
        mavenLocal()
        // used to download antlr-kotlin gradle plugin
        gradlePluginPortal()
        // used to download antlr-kotlin-gradle-plugin
        maven("https://jitpack.io")
    }
}
```

If this is still not working, please try this two files:
https://github.com/drieks/antlr-kotlin/blob/master/antlr-kotlin-examples-mpp/build.gradle.kts
https://github.com/drieks/antlr-kotlin/blob/master/antlr-kotlin-examples-mpp/settings.gradle.kts

**Comment by Lakedaemon** [comment]
thanks for your very quick and helpful answer

With your help, after investigating/setting up the whole antlr-kotlin project, experimenting a bit, I was able to understand what did not worK : 

The ide was flagging some part of the code in red, which fooled me : 
the code was actually working (trying to find something wrong in code that works as expected is one of the most difficult thing to do)

When I realized that (after 3 hours :/ ), I started to adapt the code in the gradle.build.kts to my needs (pacjage, directory, visitors, ...)^and I managed to get a multiplatform parser.

There were 3 issues with the generated parser though : 
1) my grammar file had a "Int" symbol that some enum class didin't like at all.
I changed the symbol name to "IntCollision" and the error went away
2) my previous antler was using "symbol()" methods names and all those had to be renamed 
"findSymbol()"
3) in my grammar, I'm adding a string field named "op" and in the generated parser, 
a "val op" was generated, which produced errors because it's value was set/changed later. 
I changed it into a var op:String=""  and the produced parser was orking.

Thanks for the help, again. :)
